image.py

```diff
@@ -0,0 +1,355 @@
+import io
+from shlex import split
+from subprocess import CalledProcessError, run
+from sys import stdin
+from typing import Union
+
+from exceptions import CommandNotFoundOnImageError
+
+
+class Image:
+    """A class which holds a reference to a given docker image and provides
+    an interface by which to interact with that image.
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
+
+    NOTE: This class does not work correctly if the image built does not have
+        bash installed.
+    """
+
+    def __init__(self, name_or_id: str):
+        """Initialize this Image. Connects the class with its ID.
+
+        Parameters
+        ----------
+        name_or_id : str
+            A name or ID by which to find this image using docker inspect.
+        """
+        self._id = get_image_id(name_or_id)
+
+    @classmethod
+    def build_from_dockerfile(
+        cls,
+        context: str,
+        tag: str,
```

Comment:
Let's make `tag` optional in order to support the ability to create untagged images.

Imagine that you wanted to create an ad hoc image to run a single command before immediately cleaning it up. You probably wouldn't care what the image was named, so it would be inconvenient to be required to supply a tag here.

```suggestion
        tag: Optional[str] = None,
```

Reply:
Hmm, I suppose the downside of this is that we'd probably need to parse the STDOUT of the `docker build` command in order to find out the ID of the created image, which is probably more error-prone than just calling `docker inspect -f={{.Id}} <tag>` afterwards.

Still, it would nice to make `tag` optional. We should consider whether or not it's worth the additional complexity.